### PR TITLE
Enable running the e2e tests in parallel groups

### PR DESCRIPTION
**Pull request by johnbelamaric**

### CODE REVIEW

**Comment by johnbelamaric** [comment]
@henderiw mentioned on a call that he runs some things in parallel. I think we can do that in the general test runs. This implements it.

/hold

**Comment by johnbelamaric** [comment]
Rebased

**Comment by johnbelamaric** [comment]
/hold cancel

This is working now. We can try updating the Prow job to execute with the parallel mode.

**Comment by electrocucaracha** [comment]
Maybe we have to set a meeting to discuss this. Testing things in parallel have several considerations about dependencies and log messages. Another thing to consider is to isolate every test, having a `setup` and `teardown` methods can be useful.

This [metric](https://github.com/nephio-project/test-infra/blob/main/e2e/lib/testing.sh#L28) can help us to prioritize some tests

**Comment by johnbelamaric** [comment]
Yeah, understood. Just trying to cut the time for e2e. I think this cuts it pretty substantially (maybe ~50%).

**Comment by johnbelamaric** [comment]
> Yeah, understood. Just trying to cut the time for e2e. I think this cuts it pretty substantially (maybe ~50%).

By the way, this PR does not enable this by default - you still have to call the tests with a special parameter to get parallel behavior.

**Comment by electrocucaracha** [comment]
> > Yeah, understood. Just trying to cut the time for e2e. I think this cuts it pretty substantially (maybe ~50%).
> 
> By the way, this PR does not enable this by default - you still have to call the tests with a special parameter to get parallel behavior.

That's fine, maybe we should start collecting some metrics to have better understanding of where to improve.

**Comment by johnbelamaric** [comment]
> BTW, it's amazing to finally see e2e in green

Agreed!!!

**Comment by johnbelamaric** [comment]
rebased

**Comment by electrocucaracha** [comment]
This is a quick overview of the e2e execution

| Action                                            | Time     |
|---------------------------------------------------|----------|
| Sandbox creation                                  | 759 secs | 
| Deploy regional workload cluster                  | 235 secs |
| Deploy edge clusters                              | 730 secs |
| Deploy free5gc-cp to regional workload cluster    | 112 secs |
| Deploy free5gc operator to all workload clusters  | 152 secs |
| Deploy free5gc UPF to edge clusters               | 358 secs |
| Deploy free5gc AMF and SMF to regional clusters   | 307 secs |
| Deploy ueransim to edge clusters                  | 371 secs |
| Vertically Scale free5gc UPF in Edge Clusters     | 291 secs |
| Vertically Scale free5gc SMF in Regional Clusters | 26 secs  |

**Comment by johnbelamaric** [comment]
Last couple failures have been this flake: `Internal error occurred: error applying patch: conflict: fragment line does not match src line`. I *think* https://github.com/GoogleContainerTools/kpt/pull/3993 will fix this issue. I am running Nephio e2e tests with that build of Porch now.

**Comment by johnbelamaric** [comment]
@electrocucaracha taking your table and mapping it to the groups:

| Group | Action                                            | Time     |
| ----- |---------------------------------------------------|----------|
| n/a | Sandbox creation                                  | 759 secs | 
| A | Deploy regional workload cluster                  | 235 secs |
| A | Deploy edge clusters                              | 730 secs |
| B | Deploy free5gc-cp to regional workload cluster    | 112 secs |
| B | Deploy free5gc operator to all workload clusters  | 152 secs |
| B | Deploy free5gc UPF to edge clusters               | 358 secs |
| B | Deploy free5gc AMF and SMF to regional clusters   | 307 secs |
| C | Deploy ueransim to edge clusters                  | 371 secs |
| D | Vertically Scale free5gc UPF in Edge Clusters     | 291 secs |
| E | Vertically Scale free5gc SMF in Regional Clusters | 26 secs  |

Roughly speaking, we should expect the times for each group to be the same as the max test time for the group. So in Group A we save 235 seconds, in Group B we save 112 + 152 + 307 = 571 seconds for a total savings of just over 10 minutes.

**Comment by electrocucaracha** [comment]
> Roughly speaking, we should expect the times for each group to be the same as the max test time for the group. So in Group A we save 235 seconds, in Group B we save 112 + 152 + 307 = 571 seconds for a total savings of just over 10 minutes.

Makes sense, I'm trying to get some Ansible metrics to understand which task is consuming most of time

**Comment by johnbelamaric** [comment]
rebased

File: e2e/lib/testing.sh [hunk 1]
```diff
@@ -21,12 +21,38 @@ function testing_run_test {
...
+        fi
+    done
+    wait
+    echo "+++++ $(date): finished test group $tg in $testdir"
+    printf "Test Group Time $tg: %s secs\n" "$(($(date +%s) - int_start))"
+
+    if [[ ${DEBUG:-false} == "true" ]]; then
+        echo "Porch Controller logs"
+        kubectl logs deployment/porch-controllers -n porch-system --since "$(($(date +%s) - int_start))s" | sed -e '/PackageVariant/!d;/resources changed/!d'
+    fi
```
Comment (electrocucaracha) [inline e2e/lib/testing.sh:57]: Maybe this code can be moved into a function

File: e2e/provision/init.sh [hunk 2]
```diff
@@ -106,4 +106,6 @@ printf "%s secs\n" "$(($(date +%s) - int_start))"
 
 if [[ $RUN_E2E == "true" ]]; then
     runuser -u "$NEPHIO_USER" "$REPO_DIR/e2e/e2e.sh"
+elif [[ $RUN_E2E == "parallel" ]]; then
+    runuser -u "$NEPHIO_USER" "$REPO_DIR/e2e/e2e.sh" "true"
 fi
```
Comment (electrocucaracha) [inline e2e/provision/init.sh:111]: What about another env variable?

```suggestion
if [[ $RUN_E2E == "true" ]]; then
    runuser -u "$NEPHIO_USER" "$REPO_DIR/e2e/e2e.sh" "${RUN_E2E_PARALLEL:-false}"
fi
```

File: e2e/lib/testing.sh [hunk 3]
```diff
@@ -21,12 +21,38 @@ function testing_run_test {
...
+    local tg=$2
+
+    local int_start=$(date +%s)
+    echo "+++++ $(date): starting test group $tg in $testdir"
+    for t in $TESTDIR/$tg-*.sh; do
+        if [[ $t == *-post.sh ]]; then
+            wait
+            testing_run_test "$t"
+        else
+            testing_run_test "$t" &
```
Comment (electrocucaracha) [inline e2e/lib/testing.sh:47]: We have to keep printing things in the console, otherwise terraform won't be notified.

**Review by electrocucaracha — COMMENTED** [review]
@johnbelamaric maybe we can split this in two PRs, the first one implementing groups functionality and the second to run things in parallel. BTW, maybe we can create multiple CI instances to run different e2e groups.

**Comment by johnbelamaric** [comment]
Not sure this is worth the complexity. Going to close it, we can revisit if we change think it will bring a big benefit.

/close

**Pull request closed without merging**